src/sql/src/pure.rs

```diff
@@ -1423,6 +1422,18 @@ async fn purify_create_table_from_source(
     // When kafka is implemented, likely just the topic-name will be used.
     let purified_export = match desc.connection {
         GenericSourceConnection::Postgres(pg_source_connection) => {
+            let requested_reference = match external_reference {
+                // The requested external reference for this table which will be resolved below to a
+                // fully-qualified / normalized reference for each source type.
+                Some(reference) => ExternalReferenceExport {
+                    reference: reference.clone(),
+                    alias: None,
+                },
+                None => sql_bail!(
+                    "An external reference is required for {} sources",
+                    connection_name
+                ),
+            };
```

Comment:
this and the identical change in the MySQL case below are just to accommodate the fact that the external-reference is now optional on the `CREATE TABLE FROM SOURCE` statement but required for these 2 source types

Reply:
I see, this is introducing source specific behavior for how external references work which we should avoid. I think what we want to do is always take the list of available references from the connection and then try to see if we have an unambiguous match. There shouldn't be anything source specific in this process. This means that a pg source that exposes a single table should happily accept an empty reference.

It's hard to manually ensure that we don't accidentally introduce source-specific behavior so when the refactor dust starts to settle we should invest some time putting this behavior behind traits so that we are forced to create a model for sources.

Reply:
>We only retrieve the upstream information for the table that is explicitly requested in the external-reference for postgres and mysql sources right now. 

> we should consider the latency and memory implications of adding a 'fetch all upstream schemas, tables, and their table-schemas' step to CREATE TABLE .. FROM SOURCE purification just so we can down-select to a single table immediately afterwards behind the cleaner abstraction.

That is a good point and the interface could definitely be the other way around. Specifically that each source implements a trait that includes a method like this:

```rust
async fn resolve_external_reference(&self, reference: Option<ExternalReference>) -> Self::Metadata;
```

The main point is to structure the code in such a way that source-specific behavior is hard to express. Having source specific behavior has bitten us countless times and is the reason why source planning is the rube goldberg machine that it is. So whenever I see it I'm worried that it might add yet another hitch to this gordian knot.

>  this work due to the end-of-september deadline

There should be some flexibility on that but I hear you. I know I'm repeating myself with these comments but it's only because I'm worried about future complications. We can proceed by carefully having these constraints in our minds but I would like to find the time and space to contain these source abstractions. It can happen at a later point but it will need our discipline to not deprioritize because there is a new thing with a new deadline that takes priority.

Reply:
Thinking a bit more about it, I think we'll have to (at least for now) fetch all upstream sources on `CREATE TABLE FROM SOURCE` in order to support the empty reference, even if the interface is not refactor behind a clean trait.

Reply:
> Thinking a bit more about it, I think we'll have to (at least for now) fetch all upstream sources on CREATE TABLE FROM SOURCE in order to support the empty reference, even if the interface is not refactor behind a clean trait.

If we want to submit the empty reference for postgres and mysql, yes we will. I plan to address this in a future PR, but based on our discussion here and in slack, I think your proposal for something similar to:
> `async fn resolve_external_reference(&self, reference: Option<ExternalReference>) -> Self::Metadata;`

makes sense.

And in each source implementation, we can choose whether we need to retrieve 'all' upstream references to disambiguate the reference, or can just retrieve the singular upstream table info if the reference is already qualified. This means that we'll only pay the additional latency cost of having to retrieve all tables if the user chooses to provide a less qualified reference or omit the reference entirely.